File: apps/agent/src/device.ts
import { createHash } from "node:crypto";
import { hostname, platform } from "node:os";

import type { AgentConfig } from "./types.js";

export function buildDeviceRegistration(config: AgentConfig) {
  const host = hostname();
  const osPlatform = platform();
  const fingerprint = createHash("sha256")
    .update(`${config.deviceId}:${host}:${osPlatform}`)
    .digest("hex");

  return {
    deviceId: config.deviceId,
    name: config.deviceName,
    hostname: host,
    platform: osPlatform,
    fingerprint,
  };
}

File: apps/agent/internal/agent/device.go
package agent

import (
	"crypto/sha256"
	"encoding/hex"
	"os"
	"runtime"
)

func buildDeviceRegistration(config AgentConfig) DeviceRegistration {
	host, err := os.Hostname()
	if err != nil {
		host = "unknown-host"
	}

	sum := sha256.Sum256([]byte(config.DeviceID + ":" + host + ":" + runtime.GOOS))

	return DeviceRegistration{
		DeviceID:    config.DeviceID,
		Name:        config.DeviceName,
		Hostname:    host,
		Platform:    runtime.GOOS,
		Fingerprint: hex.EncodeToString(sum[:]),
	}
}
